proposals/0000-forced-class-instantiation.rst

```diff
@@ -0,0 +1,142 @@
+.. proposal-number::
+.. trac-ticket::
+.. implemented::
+.. highlight:: haskell
+
+This is a GHC proposal `under discussion`_.
+
+.. _`under discussion`: https://github.com/ghc-proposals/ghc-proposals/pull/23
+
+Forced Class Instantiation
+==========================
+
+A declaration
+::
+  instance force Cls Ty
+
+instantiates any ``Cls a`` constraint encountered, including in error messages, otherwise ambiguous types.
+
+Motivation
+----------
+
+I am teaching a Haskell class right now, and in the first few lectures, I had to tell them: „Whenever you see ``Foldable t => … t a …`` anywhere, just mentally replace this with ``… [a] …``.”. Similarly, when discussing ``IO`` before discussing `Monad`, I had to tell them the same for these two. Finally, the first few weeks (because of using CodeWorld_), we turned on ``OverloadedStrings``, but I really do not want want to see them `IsString a` and rather have them see ``Text`` in error messages.
+
+.. _CodeWorld: http://code.world/haskell
+
+The students coped, but it would be better if the compiler did that for them. So the main motivation is to tell them to include a line like
+::
+  instance force Foldable []
+
+in their file (students accept instructions to include stuff they do not understand yet, so this is fine) and get a lower entry level.
+
+But it is useful not only for teaching. Especially
+::
+  instance IsString Text
+is useful for anyone who does not want literals overloaded but really wants them to be monomorphically the desired ``Text`` type.
+
+Proposed Change
+---------------
+I propose to introduce a declaration 
+::
+  instance force Cls Ty1 Ty2 …
+  
+The declaration is valid if the mentioned instance exists (i.e. if the constraing ``Cls Ty1 Ty2`` can be completely solved). If not, as in ``instance force Eq [a]``, a usual error message (“Could not solve constraint Eq [a]: No instance [a]” or similar) is produced, and the program is rejected.
```

Comment:
What precisely, do you mean by "can be completely solved"?  I think perhaps you mean that it can be solved from an empty context, perhaps taking multiple steps.   so `instance force Eq [[[[Int]]]]` would be ok.  Please say, with precision